tests/suite/known_triples.rs

```diff
@@ -0,0 +1,92 @@
+use std::collections::BTreeSet;
+
+use platforms::Platform;
+
+#[test]
+fn gen_known_triples() {
+    let output = "src/dist/triple/known.rs";
+    let existing = std::fs::read_to_string(output).unwrap();
+    assert_eq!(known_triples(), existing);
+}
+
+/// Generates the lists of known target architectures, OSes and environments.
+fn known_triples() -> String {
+    /// Parses the given triple into 3 parts (target architecture, OS and environment).
+    ///
+    /// # Discussion
+    ///
+    /// The current model of target triples in Rustup requires some non-code knowledge to correctly generate the list.
+    /// For example, the parsing results of two 2-dash triples can be different:
+    ///
+    /// ```jsonc
+    /// { arch: aarch64, os: linux, env: android }
+    /// { arch: aarch64, os: unknown-freebsd}
+    /// ```
+    ///
+    /// Thus, the following parsing scheme is used:
+    ///
+    /// ```jsonc
+    /// // for `x-y`
+    /// { arch: x, os: y }
+    ///
+    /// // special case for `x-y-w` where `y` is `none` or `linux`
+    /// // e.g. `thumbv4t-none-eabi`, `i686-linux-android`
+    /// // (should've been called `x-unknown-y-w`, but alas)
+    /// { arch: x, os: y, env: w }
+    ///
+    /// // for `x-y-z`
+    /// { arch: x, os: y-z }
+    ///
+    /// // for `x-y-z-w`
+    /// { arch: x, os: y-z, env: w }
+    /// ```
+    fn parse_triple(triple: &str) -> Option<(&str, &str, &str)> {
+        match triple.split('-').collect::<Vec<_>>()[..] {
+            [arch, os] => Some((arch, os, "")),
+            [arch, os @ ("none" | "linux"), env] => Some((arch, os, env)),
+            [arch, _, _] => Some((arch, &triple[(arch.len() + 1)..], "")),
+            [arch, _, _, env] => Some((
+                arch,
+                &triple[(arch.len() + 1)..(triple.len() - env.len() - 1)],
+                env,
+            )),
+            _ => None,
```

Comment:
Do these cases happen? Why? (Can you show an example?)

Reply:
@djc Nice catch! It'd be much better to let it crash there.